_posts/2021-10-15-quarkus-native-neo4j-call-tree.adoc

```diff
@@ -0,0 +1,243 @@
+that includes package and classname (only the first letter of each),
+and the method name in camel case with single letters.
+E.g. `j.p.EM.persist` would be the `display` for the `persist` method in `javax.persistence.EntityManager`.
+
+Let's repeat the query after modifying the browser style and moving the nodes to clearly view them:
+
+image::persist-query-big-nodes.png[]
+
+We can see above that one of the `persist` is to `javax.persistence.EntityManager`.
+This is the JPA method for persisting entities and the one we'll be exploring further.
+Let's narrow the query down to that one to have a clearer view:
+
+```
+match (m:Method) where m.name = "persist" and m.type =~ ".*EntityManager" return *
+```
+
+image::entitymanager-persist-query.png[]
+
+If we select the node,
+notice how we get information about the method itself.
+
+Going back to the original question,
+we wanted to find out why a given code path gets included.
+One way to do it is to start by the method itself,
+and then walk backwards to find what links
+(e.g. direct calls, virtual calls, overrides...etc)
+exist to that method within a certain depth.
+For example, let's try to find what other methods have a direct link to the `persist` method:
+
+```
+match (m:Method) <- [r*1..1] - (o) where m.name = "persist" and m.type =~ ".*EntityManager" return *
+```
+
+image::entitymanager-persist-depth-1-query.png[]
+
+Aha, so there's only one path and that's a virtual call (iow, an interface call) that comes from the `create` method in the `org.acme.hibernate.orm.FruitResource` class,
+which takes a `org.acme.hibernate.orm.Fruit` parameter and returns a `javax.ws.rs.core.Response`.
+
+Next, let's expand the query further and try to find all links with a depth of 2 to the `persist` method:
+
+```
+match (m:Method) <- [r*1..2] - (o) where m.name = "persist" and m.type =~ ".*EntityManager" return *
+```
+
+image::entitymanager-persist-depth-2-query.png[]
+
+image::subclass.png[]
+
+image::reflection-access-holder.png[]
+
+As we peel further back,
+we start to see some generated classes that invoke the `create` method in `org.acme.hibernate.orm.FruitResource`.
+`org.acme.hibernate.orm.FruitResource_ClientProxy` and `org.acme.hibernate.orm.FruitResource_Subclass` both directly call the method.
+One more interesting call comes the `FruitResource_create_d0...` method in `com.oracle.svm.core.reflect.ReflectionAccessorHolder`.
+This essentially means that the `create` method has been registered in GraalVM for access via reflection.
+
+If we query for a depth of 3, we'll find that the reflection access is an entry point.
+So, we've found the shortest path to the `persist` method,
+but that's not necessarily the only path:
+
+image::entitymanager-persist-depth-3-query.png[]
+
+You can continue going up layers, 
+but unfortunately if you reach a depth with too many nodes,
+the Neo4j browser will be unable  visualize them all.
+In this particular case, we've observed that this happens at depth of 6.
```

Comment:
I was able to visualize even 10 (I didn't try larger depths because it started taking longer)